gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaAvroExtractor.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.avro.io.DecoderFactory;
+
+import com.google.common.base.Preconditions;
+
+import gobblin.configuration.WorkUnitState;
+import gobblin.source.extractor.DataRecordException;
+
+
+public class KafkaAvroExtractor extends KafkaExtractor<Schema, GenericRecord> {
+
+  public static final String KAFKA_SCHEMA_REGISTRY_URL = "kafka.schema.registry.url";
+  public static final int SCHEMA_ID_LENGTH_BYTE = 16;
+  private static final byte MAGIC_BYTE = 0x0;
+
+  private final Schema schema;
+  private final KafkaAvroSchemaRegistry schemaRegistry;
+
+  public KafkaAvroExtractor(WorkUnitState state) {
+    super(state);
+    Preconditions.checkArgument(state.contains(KAFKA_SCHEMA_REGISTRY_URL), "Schema registry URL not provided.");
+    this.schemaRegistry = new KafkaAvroSchemaRegistry(state.getProp(KAFKA_SCHEMA_REGISTRY_URL));
+    this.schema = getLatestSchemaByTopic();
+  }
+
+  private Schema getLatestSchemaByTopic() {
+    try {
+      return this.schemaRegistry.getLatestSchemaByTopic(this.partition.getTopicName());
+    } catch (SchemaNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public Schema getSchema() {
+    return this.schema;
+  }
+
+  @Override
+  public GenericRecord readRecord(GenericRecord reuse) throws DataRecordException, IOException {
+    if (this.nextWatermark >= this.highWatermark) {
+      return null;
+    }
+    if (this.messageIterator == null || !this.messageIterator.hasNext()) {
+      this.messageIterator =
+          this.kafkaWrapper.fetchNextMessageBuffer(this.partition, this.nextWatermark, this.highWatermark);
+      if (this.messageIterator == null || !this.messageIterator.hasNext()) {
+        return null;
+      }
+    }
+
+    MessageAndOffset nextValidMessage = null;
+    do {
+      if (!this.messageIterator.hasNext()) {
+        return null;
+      }
+      nextValidMessage = this.messageIterator.next();
+    } while (nextValidMessage.offset() < this.nextWatermark);
+
+    this.nextWatermark = nextValidMessage.offset() + 1;
+    return decodeRecord(nextValidMessage, reuse);
+  }
+
+  @Override
+  protected GenericRecord decodeRecord(MessageAndOffset messageAndOffset, GenericRecord reuse) {
+    byte[] payload = getBytes(messageAndOffset.message().payload());
+    if (payload[0] != MAGIC_BYTE) {
+      throw new RuntimeException(String.format("Unknown magic byte for topic %s, partition %d",
+          this.partition.getTopicName(), this.partition.getId()));
+    }
+
+    byte[] schemaIdByteArray = new byte[SCHEMA_ID_LENGTH_BYTE];
+    schemaIdByteArray = Arrays.copyOfRange(payload, 1, 1 + SCHEMA_ID_LENGTH_BYTE);
+    String schemaId = byteArrayToHexString(schemaIdByteArray);
+    Schema schema = null;
+    try {
+      schema = this.schemaRegistry.getSchemaById(schemaId);
+    } catch (SchemaNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+
+    DatumReader<Record> reader = new GenericDatumReader<Record>(schema);
```

Comment:
Should reader be only instantiated once for the class? I don't know how expensive this line is to repeat for every record.